unfork glutin

paulrouget opened the pull request:
I'm taking over #19120.

Fix #18918.

Beside the change in the API, I had to rewrite main loop.

At this point, we should have a very similar behavior as servo-glutin.


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19895)
<!-- Reviewable:end -->

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

paulrouget commented:
> still not mouse coordinates on mouse events. I'm not worried about that. Can be a followup.

Apparently, this has been fixed in Winit.

From #19120

> Get the platform_window() code working [or drop CEF for now?].

I implemented it.

paulrouget commented:
test-tidy tells me:

```
./Cargo.lock:1: duplicate versions for package `gdi32-sys`
        The following packages depend on version 0.1.2 from 'crates.io':
                glutin
        The following packages depend on version 0.2.0 from 'crates.io':
                dwrote
                glutin_app
                servo-glutin
./Cargo.lock:1: duplicate versions for package `user32-sys`
        The following packages depend on version 0.1.3 from 'crates.io':
                glutin
        The following packages depend on version 0.2.0 from 'crates.io':
                clipboard-win
                glutin_app
                servo-glutin
./Cargo.lock:1: duplicate versions for package `gl_generator`
        The following packages depend on version 0.7.0 from 'crates.io':
                glutin
        The following packages depend on version 0.8.0 from 'crates.io':
                gleam
                glx
                offscreen_gl_context
                rust-webvr
                servo-glutin
```

I need to update upstream glutin to meet Servo's versions.

We need to remove servo-glutin from servo-skia.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19759) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
Beside Cargo.lock and an intermittent, it looks good.

Eijebong commented:
Parts of the Cargo.lock stuff will be fixed by https://github.com/tomaka/glutin/pull/982 (and a release). The rest can probably be ignored for now as it's related to the winapi 0.3 bump (which we're still waiting on mio to complete)

paulrouget commented:
What needs to be done:

- merge https://github.com/tomaka/glutin/pull/982
- merge https://github.com/tomaka/winit/pull/390
- update servo-skia to use glutin (updating the toml is enough)
- update servo-azure to use the updated servo-skia
- update the PR to use the updated servo-azure

For now, we're blocked on the winit and glutin PRs. I asked @tomaka to take a look.

paulrouget commented:
I have identify some more PR to upstream, mobile related:

- https://github.com/servo/glutin/pull/121
- https://github.com/servo/glutin/pull/117
- https://github.com/servo/glutin/pull/134
- https://github.com/servo/glutin/pull/133

/cc @MortimerGoro @NatalyaKovalova We are unforking glutin. I have tested only on Desktop. How can I make sure we are not regressing on Android and iOS?

paulrouget commented:
And:

- https://github.com/servo/glutin/pull/88
- https://github.com/servo/glutin/pull/86

paulrouget commented:
glutin-servo compiles for `aarch64-apple-ios`, but upstream glutin doesn't.

winit appears to compile and be functional. But I believe the glutin code has never been updated since the winit split, and still has plenty of code that should live in winit.

@tomaka, can you confirm that iOS is not supported by glutin, but supported by winit?

@larsbergstrom @metajack: can you tell me what is the current status of servo on iOS? I need to understand if I need to fix upstream glutin to support iOS, and how far I need to go.

tomaka commented:
>  can you confirm that iOS is not supported by glutin, but supported by winit?

The code exists but doesn't work IRC.

MortimerGoro commented:
@paulrouget Glutin required this PR merged upstream for a correct lifecycle on Android: https://github.com/tomaka/glutin/pull/868

I couldn´t land the PR at that time because Glutin just changed the API and I couldn´t do the EventLoop override. I can check the current API again and redo the PR if you want ;)

paulrouget commented:
> I couldn´t land the PR at that time because Glutin just changed the API and I couldn´t do the EventLoop override. I can check the current API again and redo the PR if you want ;)

I think I got that covered. Thank you. I'll ask you to look at the PR once ready.

metajack commented:
@paulrouget As far as I'm aware, we've never made a serious effort to port Servo to iOS, but we have discussed it a number of times. It is thought to be reasonably easy since it is so close to macOS, and @luser did a port for Gecko that was apparently not too hard.

paulrouget commented:
Unforking glutin will regress some iOS. glutin-servo compiled. upstream glutin doesn't.

larsbergstrom commented:
@paulrouget In the context of providing native WebXR support (VR and especially AR via ARKit), we funded the early and awesome work Natalya in order to enable it. Our current strategy (https://blog.mozvr.com/experimenting-with-ar-and-the-web-on-ios/ ) involves using a native web view and async communication of all AR & camera data which is pretty slow, laggy, and inefficient. It works well enough for demos, but we're assuming we will need a full native engine in the future for this to work well.

If we unfork to upstream glutin, we'll probably have to redo/test this iOS work though as @metajack says, we're not currently testing it. We didn't get to 100% of the port completed before the contract ran out.

paulrouget commented:
@alicemaz See https://groups.google.com/forum/#!topic/mozilla.dev.servo/5MHDnfMya3s - as for alternative, it's possible to embed Servo: https://github.com/paulrouget/servo-embedding-example#tutorial-embedding-servo

paulrouget commented:
> If we unfork to upstream glutin, we'll probably have to redo/test this iOS work

@larsbergstrom Before or after unforking?

Where can I get information about the iOS port? I'm not sure how to build and run the project.

larsbergstrom commented:
We were coordinating the work here, which also has info on how to build it:
https://github.com/servo/servo/issues/18154

I'm not horribly inclined to block the work to unfork on it, so long as we have permission to land support upstream in glutin. I'd hate to unfork and then refork if we need iOS support in a few months :-)

paulrouget commented:
Addressing the build issue here: https://github.com/servo/skia/pull/150

paulrouget commented:
Expected regressions after this lands:

- menus and icon of the app will probably won't work correctly on all platforms
- whatever we did for iOS won't work

paulrouget commented:
@MortimerGoro actually, I could use your help to finish up https://github.com/tomaka/glutin/pull/868

paulrouget commented:
Expected regressions after this lands:

Easy fix:

- menus and icon of the app will probably won't work correctly on all platforms (https://github.com/servo/glutin/pull/86 and https://github.com/servo/glutin/pull/88). These are not hard. Will happen in a follow up.

Difficult:

- whatever we did for iOS won't work (including https://github.com/servo/glutin/pull/134/files). Also, the winit split broke iOS compilation.
- android: life cycle issue (when apps goes to background). This is missing: https://github.com/tomaka/glutin/pull/868 and we don't have a good solution to this problem yet.

After unforking, I can commit to:

- fix the icon and app issue
- help make upstream glutin compile on iOS. This requires porting the previous servo-glutin PRs, but also fix the upstream glutin code that never got migrated during the winit split, and that will take time.
- work on the android life cycle issue, but this will depend on some winit work that I don't have control over: https://github.com/tomaka/winit/issues/231

Whoever reviews this needs to take into consideration these regressions, and tell me if it's a ok compromise.

Also, I don't know if I broke anything on the WebVR/Android front. Any help here would be great.

paulrouget commented:
@wafflespeanut wanna review this?

paulrouget commented:
(just the last commit needs review, other commits are being reviewed here #19975)

MortimerGoro commented:
> @MortimerGoro actually, I could use your help to finish up tomaka/glutin#868

I can try to redo the PR using the hack we talked about in https://github.com/tomaka/glutin/pull/868#issuecomment-319465440

It shouldn´t be difficult (if the API have not changed again :))

jdm commented:
This seems sensible, but I'd like @glennw to take a look at the integration between the event loop and compositing.

paulrouget commented:
@glennw I think the 2 relevant parts are:

https://github.com/servo/servo/pull/19895/commits/2e324e623eef02f93948f04d9f89403f92663cca#diff-b6c6a2a41e5bba045d828c36281150e9R184

and 

https://github.com/servo/servo/pull/19895/commits/2e324e623eef02f93948f04d9f89403f92663cca#diff-4d1fd4a873c7f3f5e352fad96ac6db87R537

paulrouget commented:
@MortimerGoro thank you.

glennw commented:
The two diffs you pointed to look reasonable, although I'm not super up to date with how current upstream glutin works in those areas.

Does the resizing work correctly on mac with this change (mac is the platform that implements a nested event loop for resize, I think)?

The animation / compositing bit looks correct to me.

I'm unsure why we have an ```Arc<Mutex<...>>``` for the event loop though? Why is that needed?

glennw commented:
I'm not sure even a ```RefCell``` is needed?

paulrouget commented:
I need a mutable reference to the event loop. And the window object (which owns the event loop) is wrapped in a Rc. And if I understand correctly, I can't get a mutable reference of an object wrapped in a Rc. So I assumed I can't do without a refcell.

glennw commented:
Ah, I forgot that we have Rc in places in the window implementation. In the future, we should see if we actually need that anymore. But for now, RefCell should be fine for the purposes of this PR :)

paulrouget commented:
@larsbergstrom I might have broken some WebVR things. See https://github.com/servo/servo/pull/19895#issuecomment-364148977

Can you please help?

larsbergstrom commented:
@paulrouget Please verify with @MortimerGoro, who can make sure WebVR things are working properly still.

MortimerGoro commented:
@paulrouget PR submitted to winit: https://github.com/tomaka/winit/pull/406. When this is merged we can land https://github.com/tomaka/glutin/pull/868 (PR updated but we need to include the new winit version here)

I'll test that WebVR is ok with the new glutin ;)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20022) made this pull request unmergeable. Please resolve the merge conflicts.

atouchet commented:
Will this PR fix #19901?

paulrouget commented:
I don't know.

paulrouget commented:
I had to change the way the glutin event loop runs. Initially, we would do `events_loop.borrow_mut().run_forever()`, making `events_loop` not accessible by the inner callback, throwing a `"already mutably borrowed: BorrowError"` error when accessing it (in the `get_screen` function for example).

See last commit. We leave `run_forever` immediately. I don't know if there's any cost of doing that though.

paulrouget commented:
Latest commit breaks live resize.

paulrouget committed:
do not use events_loop during run_forever

paulrouget commented:
This breaks vertical resize.

paulrouget committed:
Resize buffer on window resize

paulrouget commented:
Found another regression: after going fullscreen, beachball.

paulrouget commented:
@glennw `repaint_synchronously()` freezes after the window goes fullscreen. I see a comment that says:

```
Repaints and recomposites synchronously. You must be careful when calling this, as if a
paint is not scheduled the compositor will hang forever
```

Which is what's happening I guess.

Any idea what might go wrong in this case?

paulrouget commented:
Here is my guess. We get a resize event during the fullscreen animation, but the size of the winit window is not different just yet, and the compositor skip resize, and a paint is not scheduled.

paulrouget commented:
Confirmed.

paulrouget committed:
Store inner_size

In Winit, the Resize event has multiple purposes (actual resize,
screen change, workspace change, …). So we might get a Resize
event even if the window size hasn't changed yet. For example,
on MacOS, when going fullscreen.

Servo assume a resize and request a sync paint, which will freeze the
app. See:

https://github.com/servo/servo/blob/0fa324872330715140c86e5125640ef23e8b072b/components/compositing/compositor.rs#L1485

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20067) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@MortimerGoro did you get a chance to test WebVR?

MortimerGoro commented:
@paulrouget I tried today but there is a Servo crash on Android (also on master)

```
02-26 18:52:16.438 20279 20320 F libc    : Fatal signal 4 (SIGILL), code 1, fault addr 0xd88c23aa in tid 20320 (LayoutThread Pi)
02-26 18:52:16.440   560   560 W         : debuggerd: handling request: pid=20279 uid=10389 gid=10389 tid=20320
02-26 18:52:16.535 20347 20347 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
02-26 18:52:16.535 20347 20347 F DEBUG   : Build fingerprint: 'google/sailfish/sailfish:7.1.2/NJH47F/4146041:user/release-keys'
02-26 18:52:16.535 20347 20347 F DEBUG   : Revision: '0'
02-26 18:52:16.535 20347 20347 F DEBUG   : ABI: 'arm'
02-26 18:52:16.536 20347 20347 F DEBUG   : pid: 20279, tid: 20320, name: LayoutThread Pi  >>> com.mozilla.servo <<<
02-26 18:52:16.536 20347 20347 F DEBUG   : signal 4 (SIGILL), code 1 (ILL_ILLOPC), fault addr 0xd88c23aa
02-26 18:52:16.536 20347 20347 F DEBUG   :     r0 00000000  r1 d8d11a55  r2 00000001  r3 d88c24c1
02-26 18:52:16.536 20347 20347 F DEBUG   :     r4 cd9f99ac  r5 00800000  r6 c0000001  r7 cd9f97d8
02-26 18:52:16.536 20347 20347 F DEBUG   :     r8 00000001  r9 00000000  sl 00000000  fp cd9f9938
02-26 18:52:16.536 20347 20347 F DEBUG   :     ip cd9f9710  sp cd9f97b0  lr d88c24cb  pc d88c23aa  cpsr 40000030
02-26 18:52:16.572 20347 20347 F DEBUG   : 
02-26 18:52:16.572 20347 20347 F DEBUG   : backtrace:
02-26 18:52:16.573 20347 20347 F DEBUG   :     #00 pc 027813aa  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (_ZN12alloc_system8platform89_$LT$impl$u20$alloc..allocator..Alloc$u20$for$u20$$RF$$u27$a$u20$alloc_system..System$GT$3oom17h2ed84653f244611eE+49)
02-26 18:52:16.573 20347 20347 F DEBUG   :     #01 pc 02781373  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (_ZN64_$LT$alloc_system..System$u20$as$u20$alloc..allocator..Alloc$GT$3oom17hde3cc25712183ff5E+20)
02-26 18:52:16.573 20347 20347 F DEBUG   :     #02 pc 018a49f1  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (_ZN81_$LT$$RF$$u27$a$u20$jemallocator..Jemalloc$u20$as$u20$alloc..allocator..Alloc$GT$3oom17hc30753a9ea234bedE+22)
02-26 18:52:16.573 20347 20347 F DEBUG   :     #03 pc 018a47eb  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (__rg_oom+26)
02-26 18:52:16.573 20347 20347 F DEBUG   :     #04 pc 01353ed1  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #05 pc 01347a43  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #06 pc 0134a40f  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (_ZN65_$LT$layout..inline..InlineFlow$u20$as$u20$layout..flow..Flow$GT$17assign_block_size17hccaed8b35ac3223cE+4390)
02-26 18:52:16.573 20347 20347 F DEBUG   :     #07 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #08 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #09 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #10 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.573 20347 20347 F DEBUG   :     #11 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.574 20347 20347 F DEBUG   :     #12 pc 012e6021  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.574 20347 20347 F DEBUG   :     #13 pc 012e5fb7  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (_ZN6layout10sequential6reflow17ha63eeb57f492c959E+50)
02-26 18:52:16.574 20347 20347 F DEBUG   :     #14 pc 00913399  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.574 20347 20347 F DEBUG   :     #15 pc 0085f9b3  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.574 20347 20347 F DEBUG   :     #16 pc 0085e33b  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.574 20347 20347 F DEBUG   :     #17 pc 009144b3  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #18 pc 0085ac87  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #19 pc 00858d63  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #20 pc 008e72c5  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #21 pc 008639df  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #22 pc 0083e41b  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #23 pc 02780c41  /data/app/com.mozilla.servo-2/lib/arm/libservo.so (__rust_maybe_catch_panic+20)
02-26 18:52:16.575 20347 20347 F DEBUG   :     #24 pc 00870d49  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #25 pc 02769671  /data/app/com.mozilla.servo-2/lib/arm/libservo.so
02-26 18:52:16.575 20347 20347 F DEBUG   :     #26 pc 00047093  /system/lib/libc.so (_ZL15__pthread_startPv+22)
02-26 18:52:16.576 20347 20347 F DEBUG   :     #27 pc 00019bdd  /system/lib/libc.so (__start_thread+6)
02-26 18:52:17.066  1920  1944 I WCNSS_FILTER: ibs_msm_serial_clock_vote: vote UART CLK OFF using UART driver's ioctl()
```
I used to fix similar crashes on Servo Android but now I'm doing some GeckoView stuff and can't test Servo so often. Do you want me to take a look at this crash?

BTW we need automatic testing ASAP on Servo android to detect these kind of failures . At least only to check if servo startups correctly using a emulator.

MortimerGoro commented:
@paulrouget I was able to test WebVR on master and glutin unfork branches using https://threejs.org/examples/webvr_rollercoaster.html (that page doesn't trigger the InlineFlow crash, not related to VR or glutin unfork). Found some issues:

- The rollercoaster scene is not full screen (ok in Servo master): missing dpi or pixel ratio calculations?
- Going to background/foreground crashes sometimes (ok in Servo master)
- Can't see the enterVR button (ok in Servo master): probably related to the size issue mentioned before
- VR presentation works ok on both

paulrouget commented:
> Going to background/foreground crashes sometimes (ok in Servo master)

Will this be fixed by https://github.com/tomaka/glutin/pull/868 ?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20058) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
```
Going to background/foreground crashes sometimes (ok in Servo master)

Will this be fixed by tomaka/glutin#868 ?
```
I think that it´s a different error because the crash didn´t seem related to EGL_BAD_SURFACE (which that PR fixes). Anyway, tomorrow I'll take a deeper look.

`@MortimerGoro can write the STR for the dpi issue (I don't know how to build Servo with WebVR enabled)?
`
Instructions for a Daydream build. I tried on the Google Pixel:

- Enable WebVR preference: "dom.webvr.enabled": true,
- ./mach build --release --android --features googlevr
- ./mach package --release --android --flavor googlevr
- ./mach install --release --android
- ./mach run --android https://threejs.org/examples/webvr_rollercoaster.html (warning: the first run loads the default url sometimes after a clean APK install)
- The Three.js scene is not full screen as it should

paulrouget committed:
unfork glutin

paulrouget commented:
The duplicate dependency error will be fixed by https://github.com/tomaka/glutin/pull/992

atouchet commented:
@paulrouget how close do you expect that this PR is to being ready? I was planning on waiting for this to see if it will fix the Windows build but some of your other comments make it sound like there are still some roadblocks in the way. Would it be a better option to try to fix the Servo Glutin fork for now instead?

paulrouget commented:
@atouchet As soon as we managed to fix the Android issues. I'd like this to land this week, but realistically, I don't think this will happen before end of next week.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20140) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
@paulrouget Good news. I updated the Glutin PR to fix the background/foreground crashes https://github.com/tomaka/glutin/pull/868/files.  This [fix](https://github.com/servo/glutin/blob/servo/src/api/egl/mod.rs#L322) was missing. The problem is that 'on_surface_destroyed' Android event can arrive with some delay because multithreading communication. Because of that, swap_buffers can be called before processing 'on_surface_destroyed' event, with the native window surface already destroyed, generating different kind of intermittent crashes.

I implemented the new solution in a different way, everything is ready in the PR.

MortimerGoro commented:
@paulrouget I'll try to take a look at the scaling issue tomorrow. 

Also found a new issue. Now android requires a touch event to see something on the screen (I think that the touch events triggers the render loop start). This didn´t happen on the forked glutin. But I remember I had the same problem on Linux desktop. Is that a known issue? Why does the first render need a touch event? cc @glennw

paulrouget commented:
@MortimerGoro does the screen update only if you touch the screen or it just requires a touch event at the beginning?

paulrouget commented:
@MortimerGoro … and the touch event thing is not WebVR specific, right?

MortimerGoro commented:
`@MortimerGoro does the screen update only if you touch the screen or it just requires a touch event at the beginning?`

it just requires a touch event at the beginning

`@MortimerGoro … and the touch event thing is not WebVR specific, right?`

Correct, I tested with a three.js WebGL scene.

MortimerGoro commented:
BTW forgot to mention that the servo render loop doesn't pause when android app goes to background, which wastes battery. I fixed that in the servo fork long time ago. 

in short, this is what's missing for a correct glutin unfork on Android:
- Update glutin to fix background crashes when the PR is merged upstream.
- Research & fix the scaling issue
- Pause the render loop when the app goes to background
- Research the required touch event at the beginning to trigger the render loop

jdm commented:
I'm tracking the layout crash in https://github.com/servo/servo/issues/20149 since I can reproduce it as well.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20153) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
I can finally reproduce the scaling and the required touch event issues. Investigating…

paulrouget commented:
Scaling issue: actually, this is a broader issue where coordinates (pixel ratio, screen size, etc.) are not sent to the constellation when the browser is not ready. This is addressed in https://github.com/servo/servo/pull/20071

I will backport the fix in this PR.

paulrouget committed:
Send window size even when there is not top level browsing context yet

paulrouget commented:
Latest commit fix the scaling issue.

paulrouget commented:
For the touch event, it's not just at the beginning. When the render loop is blocked by winit (not by swap buffer), the loop is not waken up for some reason. As if `android_glue::wake_event_loop` didn't work.

paulrouget commented:
I think that the event loop is properly awaken by Android, but winit doesn't append any new event when this happens, so the event loop callback is never called. Basically, it sounds like Winit, on Android, doesn't add the `Event::Awakened` when the even loop is awaken.

paulrouget commented:
Winit PR https://github.com/tomaka/winit/pull/417 fixes the touch event issue.

paulrouget commented:
@MortimerGoro:

- Update glutin to fix background crashes when the PR is merged upstream.

Is it https://github.com/tomaka/glutin/pull/868 ?

- Research & fix the scaling issue

Fixed in this PR.

- Pause the render loop when the app goes to background

That's the job of Winit, right? I think this should work well *unless* the event loop is in `is_animating` mode. So Servo needs to know if we are backgrounded. Or could swap_buffer block until the app is back in the foreground?

- Research the required touch event at the beginning to trigger the render loop

Fixed in https://github.com/tomaka/winit/pull/417 - will need a version bump of winit, an update of glutin and of servo.

paulrouget committed:
winit and glutin versions bump (android fixes)

paulrouget commented:
@MortimerGoro about:

- Pause the render loop when the app goes to background

Did this use to work?

paulrouget commented:
Latest winit requires android_glue 2.3, and now I get this error:

E AndroidRuntime: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "cargo_apk_injected_glue_add_sync_event_handler" referenced by "/data/app/com.mozilla.servo-1/lib/arm/libservo.so"

paulrouget committed:
do not run after we get a Suspend event

paulrouget commented:
Commit https://github.com/servo/servo/pull/19895/commits/60fc3b36c17b9e92a8d846606d38900e17709429 fixes the background crash.

At that point, we good to go I believe.

This is what we need now:

1) a review of the last commits (only the first one got a r+) (@MortimerGoro can you look at the last one, where we skip the servo callback on suspend)
2) need https://github.com/tomaka/glutin/pull/1000
3) need https://github.com/servo/skia/pull/154

paulrouget commented:
@jdm r? no need to re-review the "unfork"

@MortimerGoro r? Can you look at the last commit and test to make sure you're happy with the final Android changes.

MortimerGoro commented on ports/servo/glutin_app/window.rs:
```diff
@@ -483,6 +485,14 @@ impl Window {
+            Event::Suspended(suspended) => {
+                self.suspended.set(suspended);
+                if suspended {
+                    self.set_animation_state(AnimationState::Idle);
```
Animation state is correctly set to Idle here, but it seems that it's overwritten by the compositor window.set_animation_state() call of the current tick. This starts the event loop again and wastes battery when the Android app is in background.

paulrouget replied:
@MortimerGoro can you try by adding a check to `suspended` in `fn is_animating()`. Return false if not animating or if suspended? Would that be enough?

MortimerGoro replied:
It works :)

```
fn is_animating(&self) -> bool {
    self.animation_state.get() == AnimationState::Animating && !self.suspended.get()
}
```

MortimerGoro commented:
@paulrouget I tested the new version, the scaling and touch event problems are fixed but the render loop is still live when the app goes to background. Animation state is correctly set to Idle on the suspended event, but it seems that it's overwritten by the compositor window.set_animation_state() call of the current tick. This starts the event loop again and wastes battery when the Android app is in background.

paulrouget committed:
Do not loop when animating and suspended

paulrouget commented:
@MortimerGoro see last commit.

jdm commented:
@bors-srrvo delegate+

jdm commented:
@bors-servo delegate+

bors-servo commented:
:v: @paulrouget can now approve this pull request

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20245) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4ae9f39 has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget 4ae9f3917a565bb0f4a509f1b8a1f47ddb4dbcac -->

bors-servo commented:
:hourglass: Testing commit 4ae9f3917a565bb0f4a509f1b8a1f47ddb4dbcac with merge c9f60c5a6778fe47cae85fcd88cb38257a220bb1...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10986), [arm32](http://build.servo.org/builders/arm32/builds/11225), [arm64](http://build.servo.org/builders/arm64/builds/11257), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11295), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8174), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8188), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11350), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3746), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3772), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8018), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8085), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3848), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3791), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6686)
Approved by: paulrouget
Pushing c9f60c5a6778fe47cae85fcd88cb38257a220bb1 to master...
